File: quick_sort_cormen.cpp
#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;

/*
lo e hi são inclusivos
*/
int partition(vi &a, int lo, int hi)
{
    int pivot = a[hi];
    /*
      O i vai demarcar o final da partição daqueles
      que são menores que o pivô
    */
    int i = lo - 1;

    for (int j = lo; j < hi; ++j)
    {
        if (a[j] <= pivot)
        {
            // temos que crescer a partição dos menores
            ++i;
            swap(a[i], a[j]);
        }
    }

    // coloca o pivô logo depois da partição dos menores
    swap(a[i + 1], a[hi]);
    return i + 1; // índice do pivô
}

/*
lo e hi são inclusivos
*/
void quick_sort(vi &a, int lo, int hi)
{
    if (lo < hi)
    {
        int pos;

        pos = partition(a, lo, hi);

        quick_sort(a, lo, pos - 1);
        quick_sort(a, pos + 1, hi);
    }
}

int main()
{
    int n, j;

    scanf("%d", &n);

    vector<int> a(n);
    for (int i = 0; i < n; ++i)
    {
        scanf("%d", &a[i]);
    }

    quick_sort(a, 0, n - 1);

    for (int i = 0; i < n; ++i)
    {
        printf("%d ", a[i]);
    }

    printf("\n");

    return 0;
}

File: quick_sort_cormen.py
def swap(array, a, b):
    aux = array[a]
    array[a] = array[b]
    array[b] = aux
'''    
lo e hi são inclusivos
'''
def partition(a, lo, hi):
    pivot = a[hi]
    '''
        O i vai demarcar o final da partição daqueles
        que são menores que o pivô
    '''
    i = lo - 1
    for j in range(lo, hi):
        if(a[j] <= pivot):
            '''
                Temos que crescer a partição dos menores
            '''
            i+=1
            swap(a, i, j)
    '''
        Coloca o pivô logo depois da partição dos menores
    '''
    swap(a, i+1, hi)
    return i+1 #indice do pivô

'''    
lo e hi são inclusivos
'''
def quicksort(a, lo, hi):
    if(lo < hi):
        pos = partition(a, lo, hi)
        quicksort(a, lo, pos-1)
        quicksort(a, pos+1, hi)


#input na mesma linha:
a = list(map(int, input().split()))
n = len(a)
#input com enter:
'''
n = int(input())
a = []
for i in range(0, n):
    a.append(int(input()))
'''
quicksort(a, 0, n-1)
print(a)
